hal/targets/hal/TARGET_STM/TARGET_STM32F4/storage_driver.c

```diff
@@ -0,0 +1,536 @@
+/* mbed Microcontroller Library
+ *******************************************************************************
+ * Copyright (c) 2016, STMicroelectronics
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ * 3. Neither the name of STMicroelectronics nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************
+ */
+
+#if DEVICE_STORAGE
+
+#include "Driver_Storage.h"
+#include "cmsis_nvic.h"
+#include "storage_driver_info.h"
+
+#include <string.h>
+
+#define FLASH_TIMEOUT_VALUE (50000) /* 50 seconds timeout */
+
+extern FLASH_ProcessTypeDef pFlash;
+
+static const ARM_DRIVER_VERSION version = {
+    .api = ARM_STORAGE_API_VERSION,
+    .drv = ARM_DRIVER_VERSION_MAJOR_MINOR(1,0)
+};
+
+static const ARM_STORAGE_CAPABILITIES capabilities = {
+    .asynchronous_ops = 1,
+    .erase_all        = 0,
+};
+
+static const ARM_STORAGE_INFO informations = {
+    .total_storage        = FLASH_INFO_TOTAL_STORAGE,
+    .program_unit         = 4, /* word (4 bytes) programming */
```

Comment:
`program_unit` can still be 1 byte; that allows finer granularity for updates. And from my reading of command codes in header files, it seems it is possible to issue a double-word operation, so `optimal_program_unit` could be as high as 8.